File: code/09-string-matching/java/KmpDemo.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public final class KmpDemo {
    public static void main(String[] args) {
        int[] lps = StringMatchers.buildLps("BAABABAA");
        System.out.println("lps: " + Arrays.stream(lps).mapToObj(String::valueOf).collect(Collectors.joining(",")));
        String text = "acebbceeaabceedb", pattern = "eeaab";
        List<Integer> matches = StringMatchers.kmpSearch(text, pattern);
        System.out.println("matches: " + matches.stream().map(String::valueOf).collect(Collectors.joining(",")));
    }
}

File: code/09-string-matching/python/kmp_demo.py
from string_matching import build_lps, kmp_search

if __name__ == "__main__":
    lps = build_lps("BAABABAA")
    print("lps:", ",".join(str(v) for v in lps))
    text, pattern = "acebbceeaabceedb", "eeaab"
    matches = kmp_search(text, pattern)
    print("matches:", ",".join(str(s) for s in matches))
